lib/grpc/server.ex

```diff
@@ -107,16 +107,23 @@ defmodule GRPC.Server do
          %{request_mod: req_mod, codec: codec, adapter: adapter, payload: payload} = stream,
          func_name
        ) do
-    {:ok, data} = adapter.read_body(payload)
 
-    case GRPC.Message.from_data(stream, data) do
-      {:ok, message} ->
-        request = codec.decode(message, req_mod)
+    try do
+      {:ok, data} = adapter.read_body(payload)
+
+      case GRPC.Message.from_data(stream, data) do
+        {:ok, message} ->
+          request = codec.decode(message, req_mod)
 
-        call_with_interceptors(res_stream, func_name, stream, request)
+          call_with_interceptors(res_stream, func_name, stream, request)
 
-      resp = {:error, _} ->
-        resp
+        resp = {:error, _} ->
+          resp
+      end
+    rescue
+      err ->
```

Comment:
I would use:

```elixir

catch
x -> ... # handle throw
 x, y -> ... # handle erlang error and raise
 ...
 end
 ```

Instead of rescue so this also captures `:erlang.error`s and `throws`:

```elixir
iex(3)> try do               
...(3)> :erlang.error("asdf")
...(3)> catch
...(3)> x, y -> {:ok, x, y}
...(3)> end
{:ok, :error, "asdf"}
iex(4)> try do               
...(4)> raise "asdf"
...(4)> catch              
...(4)> x, y -> {:ok, x, y}
...(4)> end
{:ok, :error, %RuntimeError{message: "asdf"}}
iex(5)> try do             
...(5)> throw "asdf"       
...(5)> catch
...(5)> x -> {:ok, x}
...(5)> end
{:ok, "asdf"}
```

Reply:
Other than this, the overall concept looks good. I'm not sure on the name, though, given we handle exceptions, erlang errors and throws

Reply:
Also, "sink" doesn't seem like the best suffix because it re-raises the exception (and would probably re-throw or re-execute the error)

Reply:
@freshtonic @lessless what do you think?

Reply:
I would go for the as complete set of data as possible, so suggestion to capture `:erlang.errors` and `throws` is an absolute 👍

Reply:
@freshtonic @lessless I'm wondering if this feature is still gonna be relevant given the telemetry support being added in #229. Wouldn't it be the case that one would just attach to the relevant client/server exception events and filter according to the %Stream{} struct attached in the metadata?